applications/SystemIdentificationApplication/custom_sensors/measurement_residual_response_function.cpp

```diff
@@ -175,26 +175,25 @@ void MeasurementResidualResponseFunction::CalculateDerivative(
     rResponseGradient.clear();
 
     auto& local_sensor_response_gradient = mResponseGradientList[OpenMPUtils::ThisThread()];
+
     double temp = 0.0;
     for (auto& p_sensor : mpSensorsList) {
-        temp += ( std::pow( p_sensor->GetValue(SENSOR_ERROR) * 0.5 * p_sensor->GetWeight(), mPCoefficient ) );
+        temp += ( std::pow( p_sensor->GetValue(SENSOR_ERROR) * p_sensor->GetWeight(), mPCoefficient ) );
     }
     const double c1 = 1 / mPCoefficient * std::pow( temp, 1/mPCoefficient - 1 );
```

Comment:
I would calculate C1, C2 in the CalculateValue and store them in private member variables of the class. Otherwise, this method is called once per element, and this C1 and C2 is calculated for every element (which is same for all the elements) which is expensive.

We always guarantee that gradients are computed if required after the CalculateValue.